src/peers.py

```diff
@@ -0,0 +1,33 @@
+from ops.framework import Object, StoredState
+
+
+class KeepalivedPeers(Object):
+
+    state = StoredState()
+
+    def __init__(self, charm, relation_name):
+        super().__init__(charm, relation_name)
+        self._relation_name = relation_name
+        self._relation = self.framework.model.get_relation(self._relation_name)
+        self.framework.observe(charm.on.keepalived_initialized, self)
+
+    @property
+    def is_joined(self):
+        return self._relation is not None
+
+    def on_keepalived_initialized(self, event):
+        if not self.framework.model.unit.is_leader():
+            return
+        # A workaround for LP: #1859769.
+        if not self.is_joined:
```

Comment:
From what I can tell, install will never have peer relations joined. You might want to reconsider how you set that up. defer here does mean that it will get triggered again by relation-created (once we add support for it) or at start() (which in my testing w/ juju 2.8 does show it as started)

```
unit-uo2-3: 17:03:27 INFO juju.worker.uniter found queued "install" hook
unit-uo2-3: 17:03:27 DEBUG juju.worker.uniter.operation running operation run install hook
unit-uo2-3: 17:03:27 DEBUG juju.machinelock acquire machine lock for uniter (run install hook)
unit-uo2-3: 17:03:27 DEBUG juju.machinelock machine lock acquired for uniter (run install hook)
unit-uo2-3: 17:03:27 DEBUG juju.worker.uniter.operation preparing operation "run install hook"
unit-uo2-3: 17:03:27 DEBUG juju.worker.uniter.operation executing operation "run install hook"
unit-uo2-3: 17:03:27 DEBUG juju.worker.uniter [AGENT-STATUS] executing: running install hook
unit-uo2-3: 17:03:27 DEBUG juju.worker.uniter.runner starting jujuc server  {unix @/var/lib/juju/agents/unit-uo2-3/agent.socket <nil>}
unit-uo2-3: 17:03:27 DEBUG worker.uniter.jujuc running hook tool "relation-ids"
unit-uo2-3: 17:03:27 DEBUG worker.uniter.jujuc running hook tool "juju-log"
unit-uo2-3: 17:03:27 INFO unit.uo2/3.juju-log install mypeer relation: None
unit-uo2-3: 17:03:27 INFO juju.worker.uniter.operation ran "install" hook (via explicit, bespoke hook script)
unit-uo2-3: 17:03:27 DEBUG juju.worker.uniter.operation committing operation "run install hook"
unit-uo2-3: 17:03:27 DEBUG juju.machinelock machine lock released for uniter (run install hook)
unit-uo2-3: 17:03:27 DEBUG juju.worker.uniter.operation lock released
unit-uo2-3: 17:03:27 INFO juju.worker.uniter.relation joining relation "uo2:mypeer"
unit-uo2-3: 17:03:27 INFO juju.worker.uniter.relation joined relation "uo2:mypeer"
unit-uo2-3: 17:03:27 DEBUG juju.worker.leadership uo2/3 is not uo2 leader
unit-uo2-3: 17:03:27 DEBUG juju.worker.uniter.operation running operation run relation-created (3; app: uo2) hook
unit-uo2-3: 17:03:27 DEBUG juju.machinelock acquire machine lock for uniter (run relation-created (3; app: uo2) hook)
unit-uo2-3: 17:03:27 DEBUG juju.machinelock machine lock acquired for uniter (run relation-created (3; app: uo2) hook)
unit-uo2-3: 17:03:27 DEBUG juju.worker.uniter.operation preparing operation "run relation-created (3; app: uo2) hook"
unit-uo2-3: 17:03:27 DEBUG juju.worker.uniter.operation executing operation "run relation-created (3; app: uo2) hook"
unit-uo2-3: 17:03:27 DEBUG juju.worker.uniter [AGENT-STATUS] executing: running mypeer-relation-created hook
unit-uo2-3: 17:03:27 DEBUG juju.worker.uniter.runner starting jujuc server  {unix @/var/lib/juju/agents/unit-uo2-3/agent.socket <nil>}
unit-uo2-3: 17:03:27 INFO juju.worker.uniter.operation skipped "mypeer-relation-created" hook (missing)
unit-uo2-3: 17:03:27 DEBUG juju.worker.uniter.operation committing operation "run relation-created (3; app: uo2) hook"
unit-uo2-3: 17:03:27 DEBUG juju.machinelock machine lock released for uniter (run relation-created (3; app: uo2) hook)
unit-uo2-3: 17:03:27 DEBUG juju.worker.uniter.operation lock released
unit-uo2-3: 17:03:27 DEBUG juju.worker.uniter.operation running operation run leader-settings-changed hook
unit-uo2-3: 17:03:27 DEBUG juju.machinelock acquire machine lock for uniter (run leader-settings-changed hook)
unit-uo2-3: 17:03:27 DEBUG juju.machinelock machine lock acquired for uniter (run leader-settings-changed hook)
unit-uo2-3: 17:03:27 DEBUG juju.worker.uniter.operation preparing operation "run leader-settings-changed hook"
unit-uo2-3: 17:03:27 DEBUG juju.worker.uniter.operation executing operation "run leader-settings-changed hook"
unit-uo2-3: 17:03:27 DEBUG juju.worker.uniter [AGENT-STATUS] executing: running leader-settings-changed hook
unit-uo2-3: 17:03:27 DEBUG juju.worker.uniter.runner starting jujuc server  {unix @/var/lib/juju/agents/unit-uo2-3/agent.socket <nil>}
unit-uo2-3: 17:03:28 INFO juju.worker.uniter.operation ran "leader-settings-changed" hook (via explicit, bespoke hook script)
unit-uo2-3: 17:03:28 DEBUG juju.worker.uniter.operation committing operation "run leader-settings-changed hook"
unit-uo2-3: 17:03:28 DEBUG juju.machinelock machine lock released for uniter (run leader-settings-changed hook)
unit-uo2-3: 17:03:28 DEBUG juju.worker.uniter.operation lock released
unit-uo2-3: 17:03:28 DEBUG juju.worker.uniter no operations in progress; waiting for changes
unit-uo2-3: 17:03:28 DEBUG juju.worker.uniter.operation running operation run config-changed hook
unit-uo2-3: 17:03:28 DEBUG juju.machinelock acquire machine lock for uniter (run config-changed hook)
unit-uo2-3: 17:03:28 DEBUG juju.machinelock machine lock acquired for uniter (run config-changed hook)
unit-uo2-3: 17:03:28 DEBUG juju.worker.uniter.operation preparing operation "run config-changed hook"
unit-uo2-3: 17:03:28 DEBUG juju.worker.uniter.operation executing operation "run config-changed hook"
unit-uo2-3: 17:03:28 DEBUG juju.worker.uniter [AGENT-STATUS] executing: running config-changed hook
unit-uo2-3: 17:03:28 DEBUG juju.worker.uniter.runner starting jujuc server  {unix @/var/lib/juju/agents/unit-uo2-3/agent.socket <nil>}
unit-uo2-3: 17:03:28 INFO juju.worker.uniter.operation ran "config-changed" hook (via explicit, bespoke hook script)
unit-uo2-3: 17:03:28 DEBUG juju.worker.uniter.operation committing operation "run config-changed hook"
unit-uo2-3: 17:03:28 DEBUG juju.machinelock machine lock released for uniter (run config-changed hook)
unit-uo2-3: 17:03:28 DEBUG juju.worker.uniter.operation lock released
unit-uo2-3: 17:03:28 INFO juju.worker.uniter found queued "start" hook
```